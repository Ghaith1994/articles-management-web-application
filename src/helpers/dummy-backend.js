let users = JSON.parse(localStorage.getItem('users')) || [];

let articles = JSON.parse(localStorage.getItem('articles')) || [];
    
export function configureDummyBackend() {
    let realFetch = window.fetch;
    window.fetch = function (url, opts) {
        return new Promise((resolve, reject) => {
            setTimeout(() => {

                // authenticate
                if (url.endsWith('/users/authenticate') && opts.method === 'POST') {
                    let params = JSON.parse(opts.body);

                    let filteredUsers = users.filter(user => {
                        return user.username === params.username && user.password === params.password;
                    });

                    if (filteredUsers.length) {
                        let user = filteredUsers[0];
                        let responseJson = {
                            id: user.id,
                            username: user.username,
                            firstName: user.firstName,
                            lastName: user.lastName,
                            token: 'fake-jwt-token'
                        };
                        resolve({ ok: true, text: () => Promise.resolve(JSON.stringify(responseJson)) });
                    } else {
                        reject('Username or password is incorrect');
                    }

                    return;
                }

                // register user
                if (url.endsWith('/users/register') && opts.method === 'POST') {
                    let newUser = JSON.parse(opts.body);

                    let duplicateUser = users.filter(user => { return user.username === newUser.username; }).length;
                    if (duplicateUser) {
                        reject('Username "' + newUser.username + '" is already taken');
                        return;
                    }

                    newUser.id = users.length ? Math.max(...users.map(user => user.id)) + 1 : 1;
                    users.push(newUser);
                    localStorage.setItem('users', JSON.stringify(users));

                    resolve({ ok: true, text: () => Promise.resolve() });

                    return;
                }


                // get articles
                if (url.endsWith('/articles') && opts.method === 'GET') {
                    if (opts.headers ) {
                        resolve({ ok: true, text: () => Promise.resolve(JSON.stringify(articles))});
                    } else {
                        reject('Unauthorised');
                    }

                    return;
                }

                // store article
                if (url.endsWith('/articles') && opts.method === 'POST') {
                    if (opts.headers  ) {
                        let newArticle = JSON.parse(opts.body);
                        newArticle.id = articles.length ? Math.max(...articles.map(article => article.id)) + 1 : 1;
                        articles.push(newArticle);
                        localStorage.setItem('articles', JSON.stringify(articles));

                        resolve({ ok: true, text: () => Promise.resolve(JSON.stringify(articles))});
                    } else {
                        reject('Unauthorised');
                    }

                    return;
                }

                // delete article
                if (url.match(/\/articles\/\d+$/) && opts.method === 'DELETE') {
                    if (opts.headers ) {
                        let urlParts = url.split('/');
                        let id = parseInt(urlParts[urlParts.length - 1]);
                        for (let i = 0; i < articles.length; i++) {
                            let article = articles[i];
                            if (article.id === id) {
                                articles.splice(i, 1);
                                localStorage.setItem('articles', JSON.stringify(articles));
                                break;
                            }
                        }

                        resolve({ ok: true, text: () => Promise.resolve() });
                    } else {
                        reject('Unauthorised');
                    }

                    return;
                }

                // update article
                if (url.match(/\/articles\/\d+$/) && opts.method === 'PUT') {
                    if (opts.headers ) {
                        let urlParts = url.split('/');
                        let id = parseInt(urlParts[urlParts.length - 1]);
                        for (let i = 0; i < articles.length; i++) {
                            let article = articles[i];
                            if (article.id === id) {
                                let newArticle = JSON.parse(opts.body);
                                articles[i] = newArticle;
                                localStorage.setItem('articles', JSON.stringify(articles));
                                break;
                            }
                        }

                        resolve({ ok: true, text: () => Promise.resolve() });
                    } else {
                        reject('Unauthorised');
                    }

                    return;
                }

                realFetch(url, opts).then(response => resolve(response));

            }, 500);
        });
    }
}