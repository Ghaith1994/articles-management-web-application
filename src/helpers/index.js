export * from './dummy-backend';
export * from './history';
export * from './store';
export * from './auth-header';