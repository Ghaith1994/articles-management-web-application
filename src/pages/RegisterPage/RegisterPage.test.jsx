import {RegisterPage} from './RegisterPage'
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import reducer from '../../reducers/index';
import { createStore, applyMiddleware } from 'redux';
import { Router } from 'react-router-dom';
import { history } from '../../helpers';
import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';

test('renders correctly without crashing', () => {
    const loggerMiddleware = createLogger();

    const store = createStore(reducer, applyMiddleware(
        thunkMiddleware,
        loggerMiddleware
    ));

    const tree = renderer.create(
        <Provider store={store}>
            <Router history={history}>
                <RegisterPage />
            </Router>
        </Provider>).toJSON();

    expect(tree).toMatchSnapshot();
});