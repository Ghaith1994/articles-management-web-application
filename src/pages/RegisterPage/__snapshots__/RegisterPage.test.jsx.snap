// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly without crashing 1`] = `
<div
  className="col-md-6 col-md-offset-3"
>
  <h2>
    Register
  </h2>
  <form
    name="form"
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        htmlFor="firstName"
      >
        First Name
      </label>
      <input
        className="form-control"
        name="firstName"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="lastName"
      >
        Last Name
      </label>
      <input
        className="form-control"
        name="lastName"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="username"
      >
        Username
      </label>
      <input
        className="form-control"
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        htmlFor="password"
      >
        Password
      </label>
      <input
        className="form-control"
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <button
        className="btn btn-primary"
      >
        Register
      </button>
      <a
        className="btn btn-link"
        href="/login"
        onClick={[Function]}
      >
        Cancel
      </a>
    </div>
  </form>
</div>
`;
